### PR TITLE
Add cloud-path interop tests using google default credentials

### PR DESCRIPTION
**Pull request by yihuazhang**
Add cloudpath interop tests that use google default credentials.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: tools/run_tests/run_interop_tests.py [hunk 1]
```diff
@@ -763,15 +766,23 @@ def cloud_to_prod_jobspec(language,
...
     cmdargs = [
         '--server_host=%s' % server_host,
         '--server_host_override=%s' % server_host, '--server_port=443',
-        '--use_tls=true',
         '--test_case=%s' % test_case
     ]
+    if transport_security == 'tls':
+        transport_security_options += ['--use_tls=true']
+    elif transport_security == 'google_default'
+        transport_security_options += ['--use_google_default=true']
```
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: any reason to not just pass `--custom_credential_type=google_default_credentials`  here? I think that would allow removal of the `--use_google_default` logic within the interop client.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,654      Total (=)      1,976,654

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,775,730      Total (<)     10,775,736

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,654      Total (=)      1,976,654

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,775,733      Total (<)     10,775,735

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: tools/run_tests/run_interop_tests.py [hunk 2]
```diff
@@ -1285,8 +1297,10 @@ def aggregate_http2_results(stdout):
 
     jobs = []
     if args.cloud_to_prod:
-        if args.transport_security != 'tls':
-            print('TLS is always enabled for cloud_to_prod scenarios.')
+        if args.transport_security != 'tls' and args.transport_security != 'google_default_credentials':
```
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: nit:  this could be slightly cleaned up to `if args.transport_security not in ['tls', 'google_default_credentials']:`

File: tools/run_tests/run_interop_tests.py [hunk 3]
```diff
@@ -1312,12 +1327,15 @@ def aggregate_http2_results(stdout):
...
                         manual_cmd_log=client_manual_cmd_log,
-                        service_account_key_file=args.service_account_key_file)
+                        service_account_key_file=args.service_account_key_file,
+                        transport_security=args.transport_security)
                     jobs.append(test_job)
 
     if args.cloud_to_prod_auth:
-        if args.transport_security != 'tls':
-            print('TLS is always enabled for cloud_to_prod scenarios.')
+        if args.transport_security != 'tls' and args.transport_security != 'google_default_credentials':
```
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: same here as for above

File: test/cpp/interop/client_helper.cc [hunk 4]
```diff
@@ -88,11 +88,13 @@ std::shared_ptr<Channel> CreateChannelForTestCase(
 
   std::shared_ptr<CallCredentials> creds;
   if (test_case == "compute_engine_creds") {
-    GPR_ASSERT(FLAGS_use_tls);
+    GPR_ASSERT(FLAGS_use_tls ||
+               FLAGS_custom_credentials_type == "google_default_credentials");
```
Comment (apolcyn) [inline test/cpp/interop/client_helper.cc]: This is a little hard to follow, because when we're using "google default credentials", we set `creds = GoogleComputeEngineCredentials();` but then don't actually use those call `creds` (if I'm correct).

Can we only set `creds = GoogleComputeEngineCredentials();` in the case that we're using TLS, and leave creds null if we're using GCD?

File: test/cpp/interop/client_helper.cc [hunk 5]
```diff
@@ -88,11 +88,13 @@ std::shared_ptr<Channel> CreateChannelForTestCase(
...
     creds = GoogleComputeEngineCredentials();
     GPR_ASSERT(creds);
   } else if (test_case == "jwt_token_creds") {
-    GPR_ASSERT(FLAGS_use_tls);
+    GPR_ASSERT(FLAGS_use_tls ||
```
Comment (apolcyn) [inline test/cpp/interop/client_helper.cc]: Might be wrong, but it seems a little strange to run "jwt_token_token" creds test with "Google Default credentials". If I'm correct, when this test runs on a VM, "Google Default Credentials" is actually using "Compute engine creds" under the hood during this test, but the test still passes because the "jwt token creds" from the service account file contains essentially the same info as the "compute engine creds"? And if this is intended to test google default creds usage of service account file when provided with one, then does the environment variable pointing to the service account file need to be set?

File: test/cpp/interop/client_helper.cc (hunk 4 above)
Comment (yihuazhang) [reply test/cpp/interop/client_helper.cc]: That's a good point. There is no need to set `creds` when using "google default credentials". The code will be revised.

File: test/cpp/interop/client_helper.cc (hunk 5 above)
Comment (yihuazhang) [reply test/cpp/interop/client_helper.cc]: AFAIK, JWT is just a format of an Oauth token, which means a JWT token obtained via ` ServiceAccountJWTAccessCredentials` should be inter-operable with that obtained via `GoogleComputeEngineCredentials`. I am also thinking that the JWT and Oauth2 tokens used in `run_interop_tests.py` should correspond to the same service account. So, I suggest to run all tests `_AUTH_TEST_CASES` with google default credentials by passing null `creds` to `CreateTestChannel` API. After all, google default credentials should have the same contents as that of TLS + {Oauth2, JWT, Compute Engine Creds}.

File: tools/run_tests/run_interop_tests.py (hunk 2 above)
Comment (yihuazhang) [reply tools/run_tests/run_interop_tests.py]: Good point.

File: tools/run_tests/run_interop_tests.py (hunk 3 above)
Comment (yihuazhang) [reply tools/run_tests/run_interop_tests.py]: Good point. Will be addressed.

File: test/cpp/interop/client_helper.cc (hunk 5 above)
Comment (apolcyn) [reply test/cpp/interop/client_helper.cc]: Ok, if surface interoperability between JWT token creds and google default creds is being tested, what do you think about: in the python tests runner, if the test case is "jwt_token_creds" and the creds are "google_default_creds", then set the environment variable so that "google default creds" finds the service account file and uses that. And if the test case is is anything besides "jwt_token_creds" and the creds are "google_default_creds", then leave the environment variable unset so as to be sure to use OAuth token?

Comment (yihuazhang) [reply test/cpp/interop/client_helper.cc]: How about we do not set the environment variable if google default creds are used regardless of test cases? Since google default creds will fetch a token from metadata server if 1) env variable is not set and 2) token is not available in a pre-defined path.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by yihuazhang** [comment]
@apolcyn, all comments are addressed. PTAL

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: tools/run_tests/run_interop_tests.py [hunk 6]
```diff
@@ -724,6 +726,9 @@ def auth_options(language, test_case, service_account_key_file=None):
     key_file_arg = '--service_account_key_file=%s' % service_account_key_file
     default_account_arg = '--default_service_account=830293263384-compute@developer.gserviceaccount.com'
 
+    # When using google_default_credentials outside of cloud-to-prod, the environment variable
```
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: nit: can we change the comment to have a "TODO" init, just to make clear that the script needs an edit for that to work.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,654      Total (=)      1,976,654

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,775,734      Total (>)     10,775,728

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by yihuazhang** [comment]
@adelez, After this PR gets merged, I plan to add `grpc_interop_google_default_toprod.cfg` to kokoro/config/prod/grpc/core/master/linux/ so that it gets triggered every couple of hours, and if everything looks good, I will make it run for every pull request. How do you think?

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,980,479      Total (=)      1,980,479

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,810,797      Total (>)     10,810,791

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yihuazhang** [comment]
Friendly ping.

File: tools/internal_ci/linux/grpc_interop_google_default_toprod.cfg [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+action {
+  define_artifacts {
+    regex: "**/sponge_log.xml"
+    regex: "github/grpc/reports/**"
+  }
+}
+
+env_vars {
+  key: "RUN_TESTS_FLAGS"
+  value: "-l c++ --transport_security google_default_credentials --cloud_to_prod --cloud_to_prod_auth --prod_servers default gateway_v4 --use_docker --internal_ci -t -j 8"
```
Comment (adelez) [inline tools/internal_ci/linux/grpc_interop_google_default_toprod.cfg]: Is it possible to include this test in existing interop test instead of creating a new set of configs? By the way, you'll need corresponding Kokoro configs in order for this to run. That's why adding to the existing job is more desirable.

Comment (yihuazhang) [reply tools/internal_ci/linux/grpc_interop_google_default_toprod.cfg]: I am fine with including the test into current cloud-to-prod interop test. But the current interop test (grpc_run_interop_tests.sh) is configured to execute run_interop_tests script only once with the existing flag `RUN_TESTS_FLAGS`, and I do not know how to execute it twice without modifying grpc_run_interop_tests.sh by introducing another one-off flag.

Comment (yihuazhang) [reply tools/internal_ci/linux/grpc_interop_google_default_toprod.cfg]: Or maybe I can modify run_interop_tests script to add another job for `cloud_to_prod` and `cloud_to_prod_auth` with google default credentials. How do you think?

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: tools/run_tests/run_interop_tests.py [hunk 8]
```diff
@@ -643,7 +643,9 @@ def __str__(self):
 #TODO: Add c++ when c++ ALTS interop server is ready.
```
Comment (jiangtaoli2016) [inline tools/run_tests/run_interop_tests.py]: pls delete this TODO and TODO three lines above

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,950,932      Total (=)      1,950,932

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,669,335      Total (>)     10,669,332

 No significant differences in binary sizes


```

File: tools/run_tests/run_interop_tests.py [hunk 9]
```diff
@@ -1285,23 +1300,40 @@ def aggregate_http2_results(stdout):
...
                                 language,
                                 test_case,
                                 server_host_nickname,
                                 prod_servers[server_host_nickname],
                                 docker_image=docker_images.get(str(language)),
                                 manual_cmd_log=client_manual_cmd_log,
                                 service_account_key_file=args.
-                                service_account_key_file)
-                            jobs.append(test_job)
+                                service_account_key_file,
```
Comment (apolcyn) [inline tools/run_tests/run_interop_tests.py]: I think this works. But it now becomes a little odd that we can no longer run the script with <i>only</i> TLS, or <i>only</i> google default creds, they need to now always be ran together.

@yihuazhang @adelez I'm thinking it might end up simpler to modify `grpc_run_interop_tests.sh` to invoke `run_interop_tests.py` twice, with different credentials.

Comment (yihuazhang) [reply tools/run_tests/run_interop_tests.py]: `grpc_run_interop_tests.sh` is also invoked by other interop tests including `grpc_interop_alts` and `grpc_interop_tocloud` so we need to modify the code in a way that it also accommodates those two cases. 

Furthermore, for cloud-path test, I think it might be good idea to test TLS and google default credentials together, and the current change in the PR (i.e., `cloud_to_prod` and `cloud_to_prod_auth`) only applies to the cloud-path test.

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,950,932      Total (=)      1,950,932

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,662,144      Total (>)     10,662,138

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yihuazhang** [comment]
Known failures: #14234
